File: java-backend/Simulador.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collections;

public class Simulador {
    //métricas do simulador
    private double utilizacaoCpu = 0;
    private double trhoughput = 0;
    private int trocaContexto = 0;
    private int tempoGlobal = 0;

    //atributos de documentacao
    private String algoritmoAtual = "";
    private String caminhoArquivoLog = "";
    private String caminhoDiretorioLog = "logs/";
    private int quantidadeExecucoes = 0;

    //atributos auxiliares
    private BufferedWriter fr;
    private ArrayList<Processo> listaProcessos;
    private ArrayList<Integer> listaTemposChegada;
    private int indiceProximaChegada;

    public  Simulador(ArrayList<Processo> listaProcessosRaw, int trocaContexto, String caminhoDiretorioLog){
        this.trocaContexto = trocaContexto;
        this.caminhoDiretorioLog = caminhoDiretorioLog;
        Collections.sort(listaProcessosRaw);
        this.listaProcessos = listaProcessosRaw;
    }

    public  Simulador(ArrayList<Processo> listaProcessosRaw, int trocaContexto){
        this.trocaContexto = trocaContexto;
        Collections.sort(listaProcessosRaw);
        this.listaProcessos = new ArrayList<Processo>(listaProcessosRaw);
    }

    public void setLista(ArrayList<Processo> listaProcessosRaw){
        Collections.sort(listaProcessosRaw);
        this.listaProcessos = new ArrayList<Processo>(listaProcessosRaw);
    }
    
    public void reset(){
        tempoGlobal = 0;
        utilizacaoCpu = 0;
        trhoughput = 0;
        indiceProximaChegada = 0;
        for(Processo p : listaProcessos){
            p.resetaExecTime();
            p.prontifica();
        }
    }

    public void executa(String escolha, boolean preemptivo) throws IOException{
        tempoGlobal = 0;
        reset();
        montaListalistaTemposChegada();
        if( escolha.equals("SJF") || escolha.equals("sjf")){
            System.out.println("****** SHORTEST JOB FIRST ******");
            algoritmoAtual = "sjf" + ((preemptivo)?" preemptivo":" nao preemptivo");
            sjf(preemptivo);
        }

        else if(escolha.equals("rr") || escolha.equals("ROUND-ROBIN") || escolha.equals("round-robin")){
            System.out.println("****** ROUND ROBIN ******");
            algoritmoAtual = "roundRobin";
            roundRobin();
        }

        else {
            System.out.println("****** FIRST COME FIRST SERVICE ******");
            algoritmoAtual = "fcfs";
            fcfs();
        }
    }

    public void geraLog() throws IOException{
        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy_MM_dd_HH-mm-ss");
        caminhoArquivoLog = caminhoDiretorioLog + algoritmoAtual +  "_" + quantidadeExecucoes + "_" + dtf.format(LocalDateTime.now())  +".log";
        File file = new File(caminhoDiretorioLog);  
        if (!file.exists()) {
            file.mkdir();
        }
        fr = new BufferedWriter(new FileWriter(caminhoArquivoLog));
        fr.write("***** " + algoritmoAtual + " *****\n");
        fr.write("*** Lista de Processos ***\n");
        for(Processo p : listaProcessos){
            fr.write("------------\n");
            fr.write(p.toString());
            
        }

        //fr.write("\n*** Descrição da execução ***\n");
        //fr.write(descricaoExecucao);

        fr.write(" \n*** Estatisticas do escalonador ***\n");
        fr.write("Utilizacao da cpu: " + utilizacaoCpu + "%\n");
        fr.write("Throughput: " + trhoughput + "\n");
        fr.write("Quantidade de processos: " + listaProcessos.size() + "\n");
        fr.write("Tempo gasto para a execucao do algoritmo: " + tempoGlobal + "\n");
        fr.write("Delay de troca de contexto: " + trocaContexto + "\n");
        fr.close();
    }

    public Processo getMenorBurstTime(ArrayList<Processo> listaProntos){
        Processo menor = listaProntos.get(0);
        for(Processo p : listaProntos){
            
            int tempoRestanteMenor = menor.getBurstTime() - menor.getTempoExecucao();
            int tempoRestanteP = p.getBurstTime() - p.getTempoExecucao();

            boolean temposRestantesIguais = tempoRestanteMenor == tempoRestanteP;
            boolean temposExecucaoIguais = p.getTempoExecucao() == menor.getTempoExecucao();
            
            if(tempoRestanteMenor > tempoRestanteP){
                menor = p;   
            }
            //criterio de desmpate 1. Maior tempo de execucao
            else if(temposRestantesIguais && ! temposExecucaoIguais){
                menor = (p.getTempoExecucao() > menor.getTempoExecucao())? p: menor;
            }    
            //criterio de desempate 2. Processo que chegou primeiro
            else if(temposRestantesIguais && temposExecucaoIguais){
                menor = (p.getArrivalTime() < menor.getArrivalTime())? p: menor;                 
            } 
        }
        return menor;
    }

    private void calculaMetricas(){
        utilizacaoCpu = (calculaTempoTotalRequisitado() / (tempoGlobal)) * 100;
        trhoughput = (tempoGlobal) / listaProcessos.size(); 
    }

    private double calculaTempoTotalRequisitado(){
        double soma = 0;
        for(Processo p : listaProcessos)
            soma += p.getBurstTime();
        return soma;    
    }

    private void montaListalistaTemposChegada(){
        listaTemposChegada = new ArrayList<>();
        for (Processo p :  listaProcessos){
            listaTemposChegada.add(p.getArrivalTime());
        }   
    }

    private void fcfs(){
        ArrayList<Processo> listaProntos = new ArrayList<Processo>();
        Processo atual = listaProcessos.get(0);
        tempoGlobal = atual.getArrivalTime();
        atual.executa();
        listaTemposChegada.remove(0);
        indiceProximaChegada = 1;
        do{
            atualizaPorChegada(listaProntos);
            if(atual.getTempoExecucao() >= atual.getBurstTime()){
                atual = escalona(atual, listaProntos);
            }
            tempoGlobal++;
            atual.atualizaExecTime();
        }while(!(listaTemposChegada.size() <= 0 && atual.finalizado()));
        tempoGlobal --;//remocao do tempo excedente
        calculaMetricas();
        quantidadeExecucoes++;
    }

    private void sjf(boolean preemptivo){
        ArrayList<Processo> listaProntos = new ArrayList<Processo>();
        tempoGlobal = 0;
        atualizaPorChegada(listaProntos);
        int indicePrimeiroProcesso = listaProntos.indexOf(getMenorBurstTime(listaProntos));
        Processo atual = listaProntos.remove(indicePrimeiroProcesso);
        atual.executa();
        do{
            ArrayList<Integer> estadoAntesChegada = new ArrayList<Integer>(listaTemposChegada);
            atualizaPorChegada(listaProntos);
            boolean chegouAlguem = !(estadoAntesChegada.equals(listaTemposChegada));
            if(preemptivo && chegouAlguem){
                atual = escalona(atual, listaProntos);
            }
    
            else if(atual.getTempoExecucao() >= atual.getBurstTime()){
                atual = escalona(atual, listaProntos);
            }
            tempoGlobal++;
            atual.atualizaExecTime();
        }while(!(listaTemposChegada.size() <= 0 && atual.finalizado()));
        tempoGlobal --;//remocao do tempo excedente
        calculaMetricas();
        quantidadeExecucoes++;
    }

    private void roundRobin(){
        ArrayList<Processo> listaProntos = new ArrayList<Processo>();
        Processo atual = listaProcessos.get(0);
        tempoGlobal = atual.getArrivalTime();
        int tempoLocalExecucao = 0;
        atual.executa();
        listaTemposChegada.remove(0);
        indiceProximaChegada = 1;
        do{
            atualizaPorChegada(listaProntos);
            if(tempoLocalExecucao >= atual.getBurstTime() || tempoLocalExecucao >= atual.getQuantum()){
                atual = escalona(atual, listaProntos);
                tempoLocalExecucao = 0;
            }
            tempoGlobal++;
            tempoLocalExecucao++;
            atual.atualizaExecTime();
        }while(!(listaTemposChegada.size() <= 0 && atual.finalizado()));
        tempoGlobal --;//remocao do tempo excedente
        calculaMetricas();
        quantidadeExecucoes++;
    }

    private Processo escalona(Processo atual, ArrayList<Processo> listaProntos){
        if(atual.getTempoExecucao() < atual.getBurstTime()){
            atual.prontifica();
            listaProntos.add(atual);
        }
        else{
            atual.finaliza(tempoGlobal);
        }

        Processo proximo = (listaProntos.size() > 0)
            ? selecionaProximo(indiceProximo(listaProntos), listaProntos)
            : null;
        return (proximo == null) ? atual : proximo;
    }

    private int indiceProximo(ArrayList<Processo> listaProntos){
        if(algoritmoAtual.contains("sjf")){
            return listaProntos.indexOf(getMenorBurstTime(listaProntos));
        }
        else{
            return 0;
        }
    }

    private Processo selecionaProximo(int indice, ArrayList<Processo>listaProntos){
        if(listaProntos.size() > 0){
            Processo proximo = listaProntos.remove(indice);
            proximo.executa();
            return proximo;
        }
        return null;
    }
    
    private void atualizaPorChegada(ArrayList<Processo> listaProntos){
        while(listaTemposChegada.size() > 0 && tempoGlobal == listaTemposChegada.get(0)){
            listaProntos.add(listaProcessos.get(indiceProximaChegada));
            listaProcessos.get(indiceProximaChegada).prontifica();
            //documentaChegada(listaProcessos.get(indiceProximaChegada));
            indiceProximaChegada++;
            listaTemposChegada.remove(0);
        }
    }
}

File: ts-backend/src/Simulador.ts
import Processo from './Processo';
import fs from 'fs';

interface ConfigDTO{
    processList: Processo[];
    contextSwitch: number;
    logPath?: string;
}

export enum tipoEscalonamento{
    fcfs = 'FirstComeFirstService', //escalonamento padrão
    sjf = 'ShortestJobFirst',
    sjf_np = 'ShortestJobFirstNaoPreemptivo',
    round_robin =  'RoundRobin'
}

export default class Simulador{
    private _simulationFrames: string[];
    private _processList: Processo[];
    private _arrivalTimes: number[];
    // métricas do escalonador
    private _contextSwitch: number;
    private _cpuUsage: number;
    private _trhoughput: number;
    private _simulationTime:number;
    private _nextArrivalIndex: number;
    // atributos de log
    private _logPath: string;
    private _currentAlgorithm: tipoEscalonamento;


    constructor({processList, contextSwitch, logPath}:ConfigDTO){
        this._processList = []
        this._arrivalTimes = []
        this._currentAlgorithm = tipoEscalonamento.fcfs;
        //atribuição e ordenação da lista de processo
        this._processList = this.cloneList(processList);
        this._processList.sort((a, b) => (a.arrivalTime < b.arrivalTime) ? -1 : 1);

        this._contextSwitch = contextSwitch;
        this._cpuUsage = 0;
        this._trhoughput = 0;
        this._simulationTime = 0;
        this._logPath = './log/';
        this._nextArrivalIndex = 0;
        this._simulationFrames = [];
    }
    
    public set contextSwitch(contextSwitch: number){
        this._contextSwitch = contextSwitch;
    }

    public set processList(processList: Processo[]){
        this._processList = this.cloneList(processList);
    }

    public changeCurrentAlgorithm(algorithm:any){
        if(algorithm != undefined){
            const choice = algorithm.toLowerCase();
            if(choice == tipoEscalonamento.sjf.toLowerCase()|| choice == 'sj' || choice == 'sjf'){
                this._currentAlgorithm = tipoEscalonamento.sjf;
            }    
            else if(choice == tipoEscalonamento.sjf_np.toLowerCase() || choice == 'sjfnp' || choice == 'sjf_np' || choice == 'sjf-np'){
                this._currentAlgorithm = tipoEscalonamento.sjf_np;
            }
            else if(choice == tipoEscalonamento.round_robin.toLowerCase() || choice == 'rr'){
                this._currentAlgorithm = tipoEscalonamento.round_robin;
            }
            else{
                this._currentAlgorithm = tipoEscalonamento.fcfs;
            }
        }
    }

    public get processList(){
        return this.cloneList(this._processList);
    }

    public exec(){
        this.reset();
        this.buildArrivalTimesList();
        if(this._processList.length > 0){
            if(this._currentAlgorithm == tipoEscalonamento.fcfs){
                //first come first service
                this.fcfs();
            }
            else if(this._currentAlgorithm == tipoEscalonamento.sjf){
                this.sjf(true);
            }   

            else if(this._currentAlgorithm == tipoEscalonamento.sjf_np){
                this.sjf(false);
            }
            else{
                this.roundRobin();
            }
            this.log();
        }
    }

    private fcfs(){
        let listaProntos:Processo[];
        listaProntos = [];
        let atual = this._processList[0]
        this._simulationTime = atual.arrivalTime;
        atual.executa();
        //sempre removemos o primeiro elemento a chegar
        this._arrivalTimes.splice(0, 1);
        this._nextArrivalIndex = 1;
        do{
            this.updateByArrival(listaProntos);
            if(atual.execTime >= atual.burstTime){
                atual = this.escalonate(atual, listaProntos);
            }
            this._simulationFrames.push(this.createFrame());
            atual.atualizaExecTime();
            this._simulationTime++;
        }while(!(this._arrivalTimes.length <= 0 && atual.finalizado()));
        this._simulationTime --;//remocao do tempo excedente
        this.computeMetrics();
    }

    private sjf(preemptive: boolean){
        //implementar o sjf aqui
    }

    private roundRobin(){
        //implementar o round robin aqui
    }

    private escalonate(atual: Processo, listaProntos:Processo[]){
        if(atual.execTime < atual.burstTime){
            atual.prontifica();
            listaProntos.push(atual);
        }
        else{
            atual.finaliza(this._simulationTime);
        }

        let proximo = (listaProntos.length > 0)
            ? this.selectNextProcess(this.nextIndex(listaProntos), listaProntos)
            : null;
        return (proximo == null) ? atual : proximo;
    }

    private computeMetrics(){
        console.log(this._simulationTime);
        console.log(this.computeRequestedTime())
        this._cpuUsage = (this.computeRequestedTime() / (this._simulationTime)) * 100;
        this._trhoughput = (this._simulationTime) / this._processList.length; 
    }

    private computeRequestedTime():number{
        let sum: number;
        sum = 0;
        this._processList.forEach((p: Processo)=>{
            sum = sum + p.burstTime;
        })

        return sum;
    }

    private nextIndex(listaProntos: Processo[]):number{
        if(this._currentAlgorithm == tipoEscalonamento.sjf){
            return listaProntos.indexOf(this.getMinorBurstTime(listaProntos));
        }
        else{
            return 0;
        }
    }

    private getMinorBurstTime(processosProntos: Processo[]):Processo{
        let min = processosProntos[0];
        processosProntos.forEach((p: Processo)=>{
            const tempoRestanteMenor = min.burstTime - min.execTime;
            const tempoRestanteP = p.burstTime - p.execTime;
    
            const temposRestantesIguais = tempoRestanteMenor == tempoRestanteP;
            const temposExecucaoIguais = p.execTime == min.execTime;
                
            if(tempoRestanteMenor > tempoRestanteP){
                min = p;   
            }
            //criterio de desmpate 1. Maior tempo de execucao
            else if(temposRestantesIguais && ! temposExecucaoIguais){
                min = (p.execTime >min.execTime)? p: min;
            }    
            //criterio de desempate 2. Processo que chegou primeiro
            else if(temposRestantesIguais && temposExecucaoIguais){
                min = (p.arrivalTime < min.arrivalTime)? p: min;                 
            } 
        })
        return min;
    }
    
    private selectNextProcess(indice: number, listaProntos: Processo[]){
        this._simulationTime += this._contextSwitch; 
        if(listaProntos.length > 0){
            let next = listaProntos.splice(indice, 1)[0];
            next.executa();
            return next;
        }
        return null;
    }

    public buildArrivalTimesList(){
        this._arrivalTimes = [];
        this._processList.forEach((p) =>{
            this._arrivalTimes.push(p.arrivalTime);
        })
    }

    public updateByArrival(listaProntos: Processo[]){
        while(this._arrivalTimes.length > 0 && this._simulationTime == this._arrivalTimes[0]){
            listaProntos.push(this._processList[this._nextArrivalIndex]);
            listaProntos[this._nextArrivalIndex - 1].prontifica();
            this._nextArrivalIndex++;
            this._arrivalTimes.splice(0,1);
        }
    }

    private reset(){
        this._simulationTime = 0;
        this._cpuUsage = 0;
        this._trhoughput = 0;
        this._simulationFrames = []
        //this._nextArrivalIndex = 0;
        this._processList.forEach((p:Processo)=>{
            p.resetaExecTime();
            p.removeDaCPU();
        })
    }

    public log(){
        const now = new Date();
        const momento_execucao = `-${now.getHours()}-${now.getMinutes()}-${now.getSeconds()}`
        const filename = `${this._logPath}${this._currentAlgorithm}${momento_execucao}.log`
        const filenameJSON = `${this._logPath}${this._currentAlgorithm}${momento_execucao}.json`
        let data = "****** SIMULAÇÂO ******\n";
        data += `Troca de contexto: ${this._contextSwitch}\n`;
        data += `Tempo de simulação: ${this._simulationTime}\n`;
        data += `Algoritmo de escalonamento: ${this._currentAlgorithm}\n\n`
        data +='Lista de Processos: \n';
        this._processList.forEach(p =>{
            data += p.stringRepresentation() + '\n' 
        });
                //Verifica se não existe
        if (!fs.existsSync(this._logPath)){
            //Efetua a criação do diretório
            fs.mkdirSync(this._logPath);
        }
        fs.writeFile(filename, data, (err) => {
            if (err) throw err;
            console.log('O arquivo de log criado!');
        });
    }
    
    public setAlgorithm(algoritmo: tipoEscalonamento){
        this._currentAlgorithm = algoritmo
    }
    
    private cloneList(arr: Processo[]):Processo[]{
        let clone: Processo[]
        clone = []
        arr.forEach((p) => {
            clone.push(new Processo({pid: p.pid,arrivalTime: p.arrivalTime,burstTime:p.burstTime, quantum: p.quantum}))      
        });
        return clone;
    }

    public getProcessList(): Processo[]{
        return this.cloneList(this._processList);
    }

    //console.log(createFrame(10, listaProcessos));
    private createFrame(){
        let frame: string; 
        frame = `{
            "currentTime": ${this._simulationTime},
            "processList":[${this._processList.map(p => p.objStringRepresentation())}]
        }`
        return frame;
    }

    public logJSONRepresentation(): JSON{
        return JSON.parse(`{
            ${this.statusOBJStringRepresentation()},
            "finalState": [${this._processList.map(p => p.objStringRepresentation())}],
            "simulationFrames":[ ${ this._simulationFrames }]
        }`);
    }

    public processListJSONRepresentation():JSON{
        return JSON.parse(`{"processList": [${this._processList.map(p => p.objStringRepresentation())}]}`)
    }

    public statusJSONRepresentation(): JSON{
        return JSON.parse(`{${this.statusOBJStringRepresentation()}}`);
    }

    private statusOBJStringRepresentation(): string{
        return `
            "currentAlgorithm" : "${this._currentAlgorithm}",
            "cpuUsage" :${this._cpuUsage},
            "throughput" : ${this._trhoughput},
            "contextSwitch" : ${this._contextSwitch},
            "simulationTime": ${this._simulationTime},
            "quantProcesses": ${this._processList.length}
        `;
    }

}